internal/addrs/module.go

```diff
@@ -168,3 +171,107 @@ func (m Module) Ancestors() []Module {
 func (m Module) configMoveableSigil() {
 	// ModuleInstance is moveable
 }
+func (m Module) configRemovableSigil() {}
+
+func parseModulePrefix(traversal hcl.Traversal) (Module, hcl.Traversal, tfdiags.Diagnostics) {
+	remain := traversal
+	var module Module
+	var diags tfdiags.Diagnostics
+
+	for len(remain) > 0 {
+		moduleName, isValidModule, moduleNameDiags := getModuleName(remain)
+		diags = diags.Append(moduleNameDiags)
+
+		if !isValidModule || diags.HasErrors() {
+			break
+		}
+
+		// Because this is a valid module address, we can safely assume that
+		// the first two elements are "module" and the module name
+		remain = remain[2:]
+
+		if len(remain) > 0 {
+			// We don't allow module instances as part of the module address
+			if _, ok := remain[0].(hcl.TraverseIndex); ok {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Module instance address with keys is not allowed",
+					Detail:   "Module address cannot be a module instance (e.g. \"module.a[0]\"), it must be a module instead (e.g. \"module.a\").",
+					Subject:  remain[0].SourceRange().Ptr(),
+				})
+			}
+		}
+
+		module = append(module, moduleName)
+	}
+
+	var retRemain hcl.Traversal
+	if len(remain) > 0 {
+		retRemain = make(hcl.Traversal, len(remain))
+		copy(retRemain, remain)
+		// The first element here might be either a TraverseRoot or a
+		// TraverseAttr, depending on whether we had a module address on the
+		// front. To make life easier for callers, we'll normalize to always
+		// start with a TraverseRoot.
+		if tt, ok := retRemain[0].(hcl.TraverseAttr); ok {
+			retRemain[0] = hcl.TraverseRoot{
+				Name:     tt.Name,
+				SrcRange: tt.SrcRange,
+			}
+		}
+	}
+
+	return module, retRemain, diags
+}
+
+func getModuleName(remain hcl.Traversal) (moduleName string, isValidModule bool, diags tfdiags.Diagnostics) {
```

Comment:
Is there any reason to return isValidModule when all callers check diags.HasErrors() ?

Reply:
Additionally, would it make sense to return the remaining traversal and remove the duplicated [2:] in both call sites?

Reply:
> Is there any reason to return isValidModule when all callers check diags.HasErrors() ?

It might just not be a module at all. If the address is `null_resource.example`, then I don't want to throw an error, but I do want to tell the caller functions they need to stop looping on the address, trying to parse the module address.
Maybe the naming here is off; if I call It `isModule` instead of `isValidModule,` will it make more sense?

> Additionally, would it make sense to return the remaining traversal and remove the duplicated [2:] in both call sites?

Makes sense. I'll do it now.

Reply:
> Additionally, would it make sense to return the remaining traversal and remove the duplicated [2:] in both call sites?

Although, on second thought, I'm not a fan of having 4 return values for this function 😬